_posts/2017-10-27-conventional-interfaces.md

```diff
@@ -0,0 +1,225 @@
+
+To demonstrate that the idea of *Sequences as Conventional Interfaces* is not limited to lists, let's show a few examples of data structures that can be thought of as sequences:
+
+* The `Map` data structure can be thought of as a generator of sequences, e.g. It can generate sequences with its keys, with its values or it can pair them together as a sequence of pairs where each element is composed of a key and a value.
+* `Maybe`, also know as `Option`, could be considered as a sequence that can contain zero or one signal.
+* `Either` as a sequence that contains a single signal which can be of one of two possible types.
+* Tuples in general and pairs `(,)` in particular could be considered as a sequence that contains exactly two signals.
+
+Different sequences have different constraints, structures and semantics, but they are equally valid from the combinators point of view. These sequences are just concrete examples of abstract mathematical structures.
+
+## Conclusion 
+
+
+By now you may have realized that both examples described in this post have a structure with the following shape, enumeration followed by a composition of standard combinators, terminating with a fold or reduction to a summary value, in other words:
+
+1. Computations always start with some sort of **enumeration**, which generates the initials signals. This is not limited to what we have seen so far, e.g. as enumerating the leaves or a tree or the integers in a given range. Enumerate could also be a query to a database that returns a sequence of products, or a sequence of products that we received via a REST endpoint. Considering lazy evaluation enumerate could also generate infinite sequences.
+2. Once we have the initial signals we pass them through the different **stages of processing**, e.g. filter, map, flatMap, zip, required to get the signals into a form in which we can extract the desired result from.
+3. To finish up the computation we **fold or reduce** the signals to a summary value.
+
+I see the pattern **Enumerate -> combinators -> Fold** as a very intuitive programming style that comes as a consequence of using **Conventional Interfaces** and it can be applied to most, if not all, computations out there. 
+
+
+To close up the post, here is a FizzBuzz implementation that follows the signal-flow structure, as an extra demonstration of this pattern.
+
+`FizzBuzz`
+
+```
+fizzBuzz :: Integer -> [String]
+fizzBuzz to = fmap fizzBuzzOrNumber (enumerateGame to)
+
+fizzBuzzOrNumber :: (String, Integer) -> String
+fizzBuzzOrNumber (fizzBuzz, index) = fizzBuzz `orElse` (show index)
+
+enumerateGame :: Integer -> [(String, Integer)]
+enumerateGame to = zip fizzesBuzzes (enumerateInterval 1 to)
+
+fizzesBuzzes :: [String]
+fizzesBuzzes = zipWith (++) fizzes buzzes
+
+fizzes :: [String]
+fizzes = cycle ["", "", "Fizz"]
+
+buzzes :: [String]
+buzzes = cycle ["", "", "", "", "Buzz"]
+
+orElse :: [a] -> [a] -> [a]
+orElse [] ys = ys
+orElse xs _  = xs
+```
+
```

Comment:
Whaaaaat, we are not using neither pygment nor rouge for syntax highlightning. We should be ashamed of ourselves. Add!!!!

Reply:
It really is, the output looks horrible :(

Reply:
I think rouge better because there is no dependency on python as with pygments. Not important on linux and macs (as they come with it out of the box), but a bit more if on Windows.